### ISSUE TITLE
Tabbed Windows

### ISSUE THREAD

**Issue by wescotta** [labwc/labwc#1932]
A few years back, it was said that the functionality for tabbed windows was going to be relegated to a utility.  Is this still the plan, or is it considerable to introduce this functionality in the near future?

There still isn't currently a functional utility native to Wayland for this functionality.  If the desire is to somehow have a utility do this, are the needed protocols, etc in place to support functionality for dragging tabs to rearrange tabs or pop another window in as a tab, among other common features?

**Comment by johanmalm** [comment]
I'm not sure what's meant by tabbed windows in this context. Could you help with a screenshot or point to another compositor or window manager that has this feature.

Apologies if I'm the guilty individual here having talking about a client (utility) implementing something.

**Comment by Consolatis** [comment]
That is something that can not be implemented by an external client (at least not without some really hacky approaches).
Technically, it could be implemented by a compositor like labwc but it feels somewhat out of scope.

**Comment by johanmalm** [comment]
Discussed on IRC. Considered out-of-scope.

**Comment by letkan** [comment]
Weren't rounded corners and shadows "out-of-scope" a while ago? Since when are functional improvements out of scope?

**Comment by johanmalm** [comment]
> Weren't rounded corners and shadows "out-of-scope" a while ago?

Rounded corners have been supported since very early versions. Possibly even before `0.1.0` - can't remember.

In the formative years we based "requirements" largely on Openbox features and categorised them here: https://github.com/labwc/labwc-scope.  I think it's fair to say that we've largely out-grown that model and should probably ponder a new model. Currently it's just down to maintainers to take a view; which has worked well for some time. 

It could be argued that at this stage it just down to some basic design ideas like:

- We don't want to grow into kwin / mutter and depend on those respective stacks
- We think that animations add more cost (through complexity and maintainability hardship) than perceived benefit
- Same of plugin architectures and custom IPC

Having said that, it's hard to set black-and-white rules on these things. Each case is considered on its own merits. For example if a team like Xfce, LXQt or RPi said that a particular feature would really make their eco-system work better and thus promotes the use of those DEs/OSs on Wayland, then I would score that very high on benefits. In any "cost" assessment it really depends on (a) the extent to which we trust the people willing to contribute code (like will they hang around and maintain it); and (b) the degree of module cohesion (e.g. if it's almost entirely in it's own translation unit - and we can largely ignore it when it comes to changes and system wide architecture).

There are two points that I can think of where we soften then original stance: icons and drop-shadows. I guess the original 'no' came from me on those two and it was mostly because I wanted to get a bunch of fundamentals right before letting focus diverge to things like parsing .desktop files.

The story with the icons was the `libsfdo` emerged - which we all liked the look of - and most maintainers were missing the titlebar application icon (top left corner), so we decided to just do it.

With drop-shadows, we had already been working (discussions and merging patches) with the Raspberry Pi team for a while. Then, one of their very competent devs opened a PR for drop-shadows. It can be a lot easier to debate these things when you have some code to ponder. As it happened - and after some lengthy discussions + reviews - everyone liked it and it got merged.

Hopefully the above gives you a good understand of how we operate - and how in fact much of FOSS world operates (in my view).

> Since when are functional improvements out of scope?

For us (and probably all other products) - scope has never been determined by the degree to which something is a functional improvement. Surely all scope relates to some 'function' - but not all functionality is every in scope on any project? :smile:
